### PR TITLE
#9987 Enable a discount for ESR payments 

### PR DESCRIPTION
**Pull request by cristinamghita**
Enable a discount for ESR payments #9987

### CODE REVIEW

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/actionhandler/impl/DiscountESRActionHandler.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		final String trxName = trxManager.getThreadInheritedTrxName(OnTrxMissingPolicy.ReturnTrxNone);
+
+//		InterfaceWrapperHelper.refresh(payment, trxName); // refresh the payment : very important; otherwise the over amount is not seen
+//		InterfaceWrapperHelper.refresh(invoice, trxName); // refresh the payment : very important; otherwise the over amount is not seen
+		Check.assumeNotNull(payment, "Null payment for line {}", line.getESR_ImportLine_ID());
+		Check.errorIf(payment.getOverUnderAmt().signum() > 0, "Exces payment for line {}. Can't discount this", line.getESR_ImportLine_ID());
+
+		final Amount discount = invoiceDAO.retrieveOpenAmt(InvoiceId.ofRepoId(invoice.getC_Invoice_ID()));
+
+		trxManager.run(trxName, new TrxRunnable()
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/actionhandler/impl/DiscountESRActionHandler.java:60]: i guess u can simply use `runInThreadInheritedTrx`

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java [hunk 2]
```diff
@@ -1746,4 +1746,40 @@ public PaymentRule getDefaultPaymentRule()
 		return sysconfigs.getReferenceListAware(SYSCONFIG_C_Invoice_PaymentRule, PaymentRule.OnCredit, PaymentRule.class);
 	}
 
+	@Override
+	public final void discountInvoice(final org.compiere.model.I_C_Invoice invoice, final BigDecimal discountAmt, final String description)
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1750]: `discount` shall be Amount.
Pls avoid using BigDecimal for amounts and quantities.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java [hunk 3]
```diff
@@ -1746,4 +1746,40 @@ public PaymentRule getDefaultPaymentRule()
...
+			// ARI
+			discountAmtAbs = discountAmt;
+		}
+
+		// @formatter:off
+		Services.get(IAllocationBL.class).newBuilder()
+			.orgId(invoice.getAD_Org_ID())
+			.currencyId(invoice.getC_Currency_ID())
+			.dateAcct(invoice.getDateAcct())
+			.dateTrx(invoice.getDateInvoiced())
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1774]: not sure if using invoice's DateAcct/DateInvoiced is correct.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java (hunk 2 above)
Comment (teosarca) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1750]: also document that `discountAmt` is not AP corrected.
i.e. "10 EUR discount" means "-10 EUR" in case of Sales Invoice and "+10 EUR" in case of Vendor invoice.

File: backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/actionhandler/impl/DiscountESRActionHandler.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		{
+			@Override
+			public void run(String trxName) throws Exception
+			{
+				// must assure that the invoice has transaction
+				InterfaceWrapperHelper.refresh(invoice, trxName);
```
Comment (teosarca) [inline backend/de.metas.payment.esr/src/main/java/de/metas/payment/esr/actionhandler/impl/DiscountESRActionHandler.java:66]: why is that needed?

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java (hunk 3 above)
Comment (cristinamghita) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1775]: Need to clear it with Mark

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java [hunk 5]
```diff
@@ -1746,4 +1747,40 @@ public PaymentRule getDefaultPaymentRule()
...
+		else
+		{
+			// ARI
+			discountAmtAbs = discountAmt.getAsBigDecimal();
+		}
+
+		// @formatter:off
+		Services.get(IAllocationBL.class).newBuilder()
+			.orgId(invoice.getAD_Org_ID())
+			.currencyId(invoice.getC_Currency_ID())
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1773]: use the discountAmt's currency

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java [hunk 6]
```diff
@@ -1746,4 +1747,40 @@ public PaymentRule getDefaultPaymentRule()
...
 
+	@Override
+	public final void discountInvoice(final @NonNull org.compiere.model.I_C_Invoice invoice, final @NonNull Amount discountAmt , final String description)
+	{
+		if (discountAmt .signum() == 0)
+		{
+			return;
+		}
+
+		final BigDecimal discountAmtAbs;
```
Comment (teosarca) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceBL.java:1758]: avoid BigDecimal at all. Use Amount type
